ambry-api/src/main/java/com/github/ambry/compression/Compression.java

```diff
@@ -0,0 +1,220 @@
+   * compressed data only.  The compressed buffer is the output of compress() that contains original/source data size.
+   * @param compressedBuffer The compressed buffer.
+   * @return Size of the original data.
+   */
+  default int getDecompressBufferSize(byte[] compressedBuffer) {
+    if (compressedBuffer == null) {
+      throw new IllegalArgumentException("compressedBuffer cannot be null.");
+    }
+    return getDecompressBufferSize(compressedBuffer, 0, compressedBuffer.length);
+  }
+
+  /**
+   * Compress portion of the shared buffer specified in {@code sourceData}.  Both {@code sourceData}
+   * and {@code compressedBuffer} can be shared/composite buffers.  That's why both buffers need to provide
+   * their size and offset.  The caller is responsible for managing memory allocation.
+   *
+   * The compressed buffer is a structure in this format:
+   * - Version (1 byte) - the structure/format version.  It starts with value '1'.
+   * - Algorithm name size (1 byte) - size of AlgorithmName in bytes.
+   * - Algorithm name (N bytes) - name of compression algorithm, up to the max size specified in CompressionFactory.
+   * - Original Data size (4 bytes) - up to 2 GB.
+   * - Compressed data - the output of the compression algorithm.
+   *
+   * Note that the compressed data does not contain original data size, but decompress() requires the original data
+   * size.  The original data size is stored as 4-bytes integer in the compressed buffer.
+   *
+   * @param sourceData The source uncompressed data to compress.  It cannot be null or empty.
+   * @param sourceDataOffset Offset in the sourceData to start compressing.
+   * @param sourceDataSize The size to compress.
+   * @param compressedBuffer The compressed buffer where the compressed data is written to.
+   * @param compressedBufferOffset Offset in compressedBuffer to write to.
+   * @param compressedBufferSize The maximum size to write inside the compressedBuffer.
+   * @return The actual compressed data size in bytes.
+   */
+  int compress(byte[] sourceData, int sourceDataOffset, int sourceDataSize,
+      byte[] compressedBuffer, int compressedBufferOffset, int compressedBufferSize);
+
+  /**
+   * Compress the source data specified in {@code sourceData}.  The source data is not a shared buffer.
+   * This method will allocate buffer to store the compressed data.
+   * The returned buffer may be bigger than sourceData in case sourceData is incompressible.
+   * The return value contains both the buffer and actual usage in the buffer.
+   *
+   * @param sourceData The source uncompressed data to compress.  It cannot be null or empty.
+   * @return Pair that contains the compressed buffer and the buffer usage size in bytes.
+   */
+  default Pair<Integer, byte[]> compress(byte[] sourceData) {
+    if (sourceData == null) {
+      throw new IllegalArgumentException("sourceData cannot be null.");
```

Comment:
the javadocs say that `sourceData` cannot be empty. Should we include an empty check too?